### PR TITLE
CORE-11066 Implement JSON factory storage and execution

### PR DESCRIPTION
**Pull request by nkovacsx**
### Overview

- Create the `ContractStateVaultJsonFactoryStorage` and `ContractStateVaultJsonFactoryProvider` to detect/register implementations of `ContractStateVaultJsonFactory`.
- Create the default internal implementation of `ContractStateVaultJsonFactoryImpl` and wire it to `ContractStateVaultJsonFactoryProvider` explicitly.
- Create the `UtxoPersistenceServiceImplTest` test class but it only tests JSON representation related logic for now.
- Create the `ContractStateVaultJsonUtilities` function library that will do the low-level JSON manipulation.

### CODE REVIEW

File: components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/utxo/tests/UtxoPersistenceServiceImplTest.kt [hunk 1]
```diff
@@ -18,7 +18,7 @@ import net.corda.ledger.persistence.utxo.CustomRepresentation
 import net.corda.ledger.persistence.utxo.UtxoPersistenceService
 import net.corda.ledger.persistence.utxo.UtxoRepository
 import net.corda.ledger.persistence.utxo.UtxoTransactionReader
-import net.corda.ledger.persistence.utxo.impl.UtxoPersistenceServiceImpl
+import net.corda.ledger.persistence.utxo.impl.UtxoPersistenceServiceImplTest
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/utxo/tests/UtxoPersistenceServiceImplTest.kt]: This change looks odd?

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/ContractStateVaultJsonFactoryStorage.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package net.corda.ledger.persistence.json
+
+import net.corda.v5.ledger.utxo.ContractState
+import net.corda.v5.ledger.utxo.query.json.ContractStateVaultJsonFactory
+
+/**
+ * A storage for contract state json factories.
+ */
+interface ContractStateVaultJsonFactoryStorage {
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/ContractStateVaultJsonFactoryStorage.kt]: Can we call this a `Registry` instead of `Storage`. Keeps it similar to what you did for `VaultNamedQueryRegistry`

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/ContractStateVaultJsonFactoryStorage.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+    /**
+     * Store a given contract state json factory.
+     */
+    fun registerJsonFactory(factory: ContractStateVaultJsonFactory<ContractState>)
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/ContractStateVaultJsonFactoryStorage.kt]: This should be `out ContractState`?

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryImpl.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        ContractStateVaultJsonFactory::class,
+        UsedByPersistence::class
+    ],
+    scope = ServiceScope.PROTOTYPE
+)
+class ContractStateVaultJsonFactoryImpl @Activate constructor()
+    : ContractStateVaultJsonFactory<ContractState>, UsedByPersistence {
+    override fun getStateType(): Class<ContractState> = ContractState::class.java
+
+    override fun create(state: ContractState, jsonMarshallingService: JsonMarshallingService): String {
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryImpl.kt:23]: Actually, when I think about it, we don't really know what we want to store for `ContractState` yet because we don't have the identity story completed.

For now, lets persist an empty string/block.

This also makes me think that if a factory returns `""` we should change it to `{}`

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryProvider.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    @Reference(service = ContractStateVaultJsonFactoryStorage::class)
+    private val factoryStorage: ContractStateVaultJsonFactoryStorage,
+
+    // The default internal implementation of `ContractStateVaultJsonFactory`
+    @Reference(service = ContractStateVaultJsonFactory::class)
+    private val internalFactory: ContractStateVaultJsonFactory<ContractState>
+) : UsedByPersistence, CustomMetadataConsumer {
+
+    private companion object {
+        private val logger = LoggerFactory.getLogger(ContractStateVaultJsonFactoryProvider::class.java)
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryProvider.kt]: DOn't need the private on the logger cos you have it on the companion object.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryProvider.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Suspendable
+    override fun accept(context: MutableSandboxGroupContext) {
+        val metadataServices = context.getMetadataServices<ContractStateVaultJsonFactory<*>>()
+
+        if (logger.isDebugEnabled) {
+            if (metadataServices.size == 1) {
+                logger.debug("Found 1 contract state vault json factory.")
+            } else {
+                logger.debug("Found ${metadataServices.size} contract state vault json factory.")
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryProvider.kt]: `factory` -> `factories`.

Comment (lankydan) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryProvider.kt]: Can we also output the discovered factories, must be useful in the future if there are some registration issues.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryStorageImpl.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        ContractStateVaultJsonFactoryStorage::class,
+        UsedByPersistence::class
+    ],
+    scope = ServiceScope.PROTOTYPE
+)
+class ContractStateVaultJsonFactoryStorageImpl @Activate constructor()
+    : ContractStateVaultJsonFactoryStorage, UsedByPersistence {
+
+    private companion object {
+        private val logger = LoggerFactory.getLogger(ContractStateVaultJsonFactoryStorageImpl::class.java)
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryStorageImpl.kt]: Double `private`.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryStorageImpl.kt [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    private val factoryStorage = ConcurrentHashMap<Class<ContractState>, ContractStateVaultJsonFactory<ContractState>>()
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryStorageImpl.kt]: `ConcurrentHashMap<Class<out ContractState>, ContractStateVaultJsonFactory<out ContractState>>()`

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryStorageImpl.kt [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    override fun registerJsonFactory(factory: ContractStateVaultJsonFactory<ContractState>) {
+        if (factoryStorage.putIfAbsent(factory.stateType, factory) != null) {
+            logger.warn("A factory for state class ${factory.stateType} is already registered.")
+            throw IllegalArgumentException("A factory for state class ${factory.stateType} is already registered.")
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryStorageImpl.kt]: Can we output the class name of the currently registered factory and the one that is trying to be registered.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryStorageImpl.kt [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        }
+    }
+
+    override fun getFactoriesForClass(state: ContractState): List<ContractStateVaultJsonFactory<ContractState>> {
+        return factoryStorage.filter {
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryStorageImpl.kt]: We may want to cache some of this in the future, but is fine for now.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/impl/VaultNamedQueryFactoryProvider.kt [hunk 11]
```diff
@@ -27,7 +27,7 @@ class VaultNamedQueryFactoryProvider @Activate constructor(
 ) : UsedByPersistence, CustomMetadataConsumer {
 
     private companion object {
-        private val logger = LoggerFactory.getLogger(this::class.java.enclosingClass)
+        private val logger = LoggerFactory.getLogger(VaultNamedQueryFactoryProvider::class.java)
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/impl/VaultNamedQueryFactoryProvider.kt]: double private

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/impl/VaultNamedQueryRegistryImpl.kt [hunk 12]
```diff
@@ -19,7 +19,7 @@ import java.util.concurrent.ConcurrentHashMap
 class VaultNamedQueryRegistryImpl @Activate constructor(): VaultNamedQueryRegistry, UsedByPersistence {
 
     private companion object {
-        private val logger = LoggerFactory.getLogger(this::class.java.enclosingClass)
+        private val logger = LoggerFactory.getLogger(VaultNamedQueryRegistryImpl::class.java)
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/impl/VaultNamedQueryRegistryImpl.kt]: double private

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt [hunk 13]
```diff
@@ -216,4 +233,28 @@ class UtxoPersistenceServiceImpl constructor(
             repository.persistTransactionStatus(em, id, transactionStatus, utcClock.instant())
         }
     }
+
+    private fun extractJsonDataFromState(
+        contractState: ContractState,
+        factoryStorage: ContractStateVaultJsonFactoryStorage,
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt]: We don't need to pass in the `factoryStorage` and `jsonMarshallingService` because they are in the services constructor args.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt [hunk 14]
```diff
@@ -216,4 +233,28 @@ class UtxoPersistenceServiceImpl constructor(
...
+        jsonMarshallingService: JsonMarshallingService
+    ): String {
+        // This block will fail with JsonProcessingException if either:
+        // - the JSON string is invalid from any of the factories
+        // - the concatenated map from the different factory outputs is invalid
+        return try {
+            val jsonMap = factoryStorage.getFactoriesForClass(contractState).associate {
+                it.stateType.name to jsonMarshallingService.parse(
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt]: Can you make sure to write a test for inner classes, which will contain a `$` in the name. I want to double check that the queries still work fine in that scenario. It probably has to be an E2E test that you'll do in your next task, but give it a quick test in your current smoke test.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt [hunk 15]
```diff
@@ -216,4 +233,28 @@ class UtxoPersistenceServiceImpl constructor(
...
+                    // We need to parse the output string to make it an actual JSON format
+                    it.create(contractState, jsonMarshallingService), Any::class.java
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt]: What happens if the output json is invalid?

File: components/ledger/ledger-persistence/src/test/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImplTest.kt [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+                "net.corda.v5.ledger.utxo.ContractState" : {
+                    "participants" : "[]" 
+                }
+            }
+            """.trimIndent())
+        )
+    }
+
+    @Test
+    fun `Persisting a transaction with no JSON factory will result in an empty JSON string being stored`() {
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/test/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImplTest.kt]: This test kind of conflicts with the test above which says it always uses the `ContractState` factory.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryProvider.kt (hunk 6 above)
Comment (blsemo) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryProvider.kt]: Can we just log `logger.debug("Number of contract state vault json factories found: ${metadataServices.size()}`.
I really dislike those if statements that change the log message depending on cardinality to make the grammar look nice.
- They look like busy code logic but do just log formatting
- They make grepping through logfiles needlessly hard.
- Logfiles aren't novels, there is no prize for nice English wording.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt [hunk 17]
```diff
@@ -216,4 +233,28 @@ class UtxoPersistenceServiceImpl constructor(
...
+                )
+            }
+
+            // Reformat the whole map as a string
+            jsonMarshallingService.format(jsonMap)
+        } catch (e: JsonProcessingException) {
+            log.warn("Error while processing JSON, reverting to empty JSON representation.")
+            "{}" // Empty JSON string if we encountered an exception
+        }
```
Comment (blsemo) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt]: I am not 100% sure this is the right approach - one bad factory can blow away all of the JSON, and all you get is a warning in the log file of the DB processor.

If that is the behaviour, users need to be made aware of that testing all the edge cases of the JSON factories is crucial in order not to end up with an unqueryable vault.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/ContractStateVaultJsonFactoryStorage.kt (hunk 2 above)
Comment (blsemo) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/ContractStateVaultJsonFactoryStorage.kt]: Agree on that one - storage sounds too much like storing data. I found it confusing when reading the rest of the PR.

File: components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/utxo/tests/UtxoPersistenceServiceImplTest.kt (hunk 1 above)
Comment (nkovacsx) [reply components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/utxo/tests/UtxoPersistenceServiceImplTest.kt]: This was probably some IntelliJ refactor mess up.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/ContractStateVaultJsonFactoryStorage.kt (hunk 2 above)
Comment (nkovacsx) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/ContractStateVaultJsonFactoryStorage.kt]: Good point

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryImpl.kt (hunk 4 above)
Comment (nkovacsx) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryImpl.kt:23]: Re. your last point, we already do that but I'll add a test case.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt (hunk 13 above)
Comment (nkovacsx) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt]: true

File: components/ledger/ledger-persistence/src/test/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImplTest.kt (hunk 16 above)
Comment (nkovacsx) [reply components/ledger/ledger-persistence/src/test/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImplTest.kt]: Well, there's a difference between absolutely no factory registered and *no specific* factory registered.
The former one means, not even `ContractState` factory is registered, the latter means no factory is registered for the given state type. 
I tried to clarify the test names.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt (hunk 17 above)
Comment (nkovacsx) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt]: I modified the logic now.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryRegistryImpl.kt [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    private val factoryStorage = ConcurrentHashMap<Class<out ContractState>, ContractStateVaultJsonFactory<out ContractState>>()
+
+    override fun registerJsonFactory(factory: ContractStateVaultJsonFactory<out ContractState>) {
+        if (factoryStorage.putIfAbsent(factory.stateType, factory) != null) {
+            logger.warn("Trying to register factory with type: $factory. " +
+                    "However, a factory with type: ${factoryStorage[factory.stateType]} is already registered " +
+                    "for state class ${factory.stateType}.")
+            throw IllegalArgumentException("A factory for state class ${factory.stateType} is already registered.")
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/json/impl/ContractStateVaultJsonFactoryRegistryImpl.kt]: Word the warning like:

```
Failed to register ${ContractStateVaultJsonFactory::class.java.name} of ${factory::class.java.name} for state type ${factory.stateType::class.java.name} as ${factoryStorage[factory.stateType]} is already registered for the same type.
```

Also change the exception message to 

```
A ${ContractStateVaultJsonFactory::class.java.name} for state class ${factory.stateType} is already registered.
```

Think it just makes things a bit more clearer.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt [hunk 19]
```diff
@@ -216,4 +230,34 @@ class UtxoPersistenceServiceImpl constructor(
...
+        }.associate {
+
+            val jsonToParse = @Suppress("unchecked_cast")
+            (it as ContractStateVaultJsonFactory<ContractState>)
+                .create(contractState, jsonMarshallingService)
+                .ifBlank { "{}" } // Default to "{}" if the provided factory returns empty string to avoid exception
+
+            it.stateType.name to try {
+                jsonMarshallingService.parse(jsonToParse, Any::class.java)
+            } catch (e: JsonProcessingException) {
```
Comment (lankydan) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt]: Honestly I think we might have to catch all `Exception`s here because we can't guarantee people don't cause weird issues leading to states/transactions not being recorded.

File: components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/utxo/tests/UtxoPersistenceServiceImplTest.kt [hunk 20]
```diff
@@ -92,6 +93,7 @@ class UtxoPersistenceServiceImplTest {
     private lateinit var entityManagerFactory: EntityManagerFactory
     private lateinit var repository: UtxoRepository
     private lateinit var cpiInfoReadService: CpiInfoReadService
+    private lateinit var factoryStorage: ContractStateVaultJsonFactoryRegistry
```
Comment (blsemo) [inline components/ledger/ledger-persistence/src/integrationTest/kotlin/net/corda/ledger/persistence/utxo/tests/UtxoPersistenceServiceImplTest.kt]: The variable should also be renamed to `factoryRegistry`.

**Review by mag1c-48 — APPROVED** [review]
LGTM - build only

**Review by lankydan — APPROVED** [review]

**Review by kyriathar — APPROVED** [review]
LGTM (db & packaging)
